kamon-akka-remote-2.5.x/src/main/scala/kamon/akka/instrumentation/cluster/ShardingInstrumentation.scala

```diff
@@ -0,0 +1,134 @@
+
+import akka.actor.{ActorRef, Terminated}
+import akka.cluster.Cluster
+import akka.cluster.sharding.ShardCoordinator.Internal.HandOff
+import akka.cluster.sharding.{Shard, ShardRegion}
+import akka.cluster.sharding.ShardRegion.{GracefulShutdown, _}
+import org.aspectj.lang.ProceedingJoinPoint
+import org.aspectj.lang.annotation._
+
+import scala.concurrent.duration._
+@Aspect
+class ShardingInstrumentation {
+
+  @Pointcut(
+    "execution(akka.cluster.sharding.ShardRegion.new(..)) && this(region) && args(typeName, *, *, *, *, extractEntityId, ..)")
+  def regionCreate(region: ShardRegion,
+                   typeName: String,
+                   extractEntityId: ShardRegion.ExtractEntityId): Unit = {}
+
+  @After("regionCreate(region, typeName, extractEntityId)")
+  def afterRegionCreate(region: ShardRegion,
+                        typeName: String,
+                        extractEntityId: ShardRegion.ExtractEntityId) = {
+    val withIdentity = region.asInstanceOf[ShardedType]
+    withIdentity.setTypeName(typeName)
+    withIdentity.setIdExtractor(extractEntityId)
+  }
+
+  @Pointcut(
+    "execution(* akka.cluster.sharding.ShardRegion.receive()) && this(region)")
+  def msgReceive(region: ShardRegion): Unit = {}
+
+  @Around("msgReceive(region)")
+  def aroungMsgReceive(pjp: ProceedingJoinPoint, region: ShardRegion) = {
+    val identifiedRegion = region.asInstanceOf[ShardedType]
+    val metrics = ShardingMetrics.forType(identifiedRegion.typeName)
+
+    def handle(msg: Any) = msg match {
+      case msg if identifiedRegion.idExtractor.isDefinedAt(msg) =>
+        metrics.regionMessage
+      case _ => ()
+    }
+
+    val receive: PartialFunction[Any, Unit] =
+      pjp.proceed().asInstanceOf[PartialFunction[Any, Unit]]
+    val wrappedReceive: PartialFunction[Any, Unit] = {
+      case msg: Any =>
+        handle(msg)
+        receive.apply(msg)
+    }
+    wrappedReceive
+  }
+
+  @Pointcut(
+    "execution(* akka.cluster.sharding.ShardRegion.postStop()) && this(region)")
+  def regionStop(region: ShardRegion): Unit = {}
+
+  @After("regionStop(region)")
+  def afterRegionStop(region: ShardRegion): Unit = {
+    ShardingMetrics.cleanInstrumentation(
+      region.asInstanceOf[ShardedType].typeName)
+  }
+
+  @Pointcut(
+    "execution(akka.cluster.sharding.Shard.new(..)) && this(shard) && args(typeName, *, *, *, extractEntityId, ..)")
+  def shardCreate(shard: Shard,
+                  typeName: String,
+                  extractEntityId: ExtractEntityId): Unit = {}
+
+  @After("shardCreate(shard, typeName, extractEntityId)")
+  def aroundShardCreate(shard: Shard,
+                        typeName: String,
+                        extractEntityId: ExtractEntityId) = {
+    val withIdentity = shard.asInstanceOf[ShardedType]
+    withIdentity.setTypeName(typeName)
+    withIdentity.setIdExtractor(extractEntityId)
+    ShardingMetrics.forType(typeName).shardStarted(shard.context.self, shard)
+
+    ()
```

Comment:
Why is this here? :thinking: I'm also seeing it on line 97 but not sure of what is the purpose. Please, add explicit return types to the functions as well :).. I know that all of these functions using `@Around` should be returning `Any` but there are a few other cases where is not clear.